ras_backstage/authentication/token_decoder.py

```diff
@@ -0,0 +1,35 @@
+import logging
+
+from flask import abort
+import jwt
+
+from ras_backstage.authentication.uaa import get_uaa_public_key
+
+
+logger = logging.getLogger(__name__)
+
+USER_ID_CLAIM = "user_id"
+
+
+def decode_access_token(access_token, verify=True):
+
+    try:
+        decoded_jwt = jwt.decode(
+            access_token,
+            verify=verify,
+            algorithms=None,
+            key=get_uaa_public_key(),
```

Comment:
I don't think we should be calling a method here to try and get the public key from the UAA. We should only need to get the public key in one place on startup and assume it's in the config after that. I also think the method should be provided with the key rather than get it from the config as this will be unit testable outside of a Flask context.

Reply:
The get_uaa_public_key method first checks if the key is in app config, if not will go and retrieve it from the UAA and put it in the config. The config is also initialised on startup but if the UAA isn't available then this allows the application to start and retrieve the key later.

def get_uaa_public_key():
    if not current_app.config.get('UAA_PUBLIC_KEY'):
        current_app.config['UAA_PUBLIC_KEY'] = request_uaa_public_key(current_app)
    return current_app.config['UAA_PUBLIC_KEY']

Reply:
I disagree that we should only be getting the key on startup. The key comes with an etag header  that allows UAA to know if the key is up to date. according to the docs, this expires on a configuration change, so could be relatively frequent compared to statically generated keys. There should be some way of refreshing that key without bringing down the service.

Reply:
That's a good point, however I would like to do that work separately. We're very unlikely to rotate the public key in the first release and if we do then an app restart is acceptable.

Reply:
https://trello.com/c/KgjFhxme/92-refresh-public-key

Reply:
@warrenbailey My concern with this is it gives a false impression that the service is running, as it will start up, but the UAA may never start. The complexity of retrying to get a key could be done by whatever is orchestrating the deployment e.g. try to redeploy every minute for x minutes. One of the current frustrations in setting up all of the services is services that have started but don't work as they are dependent on other services that haven't. 

@JamesGardiner Currently once it's got a key, it puts it in the config and doesn't call the UAA again. Future work can look at how we might refresh that key, but we'd have to consider if we want to make a request to the UAA for a key on every decode.

Reply:
A big problem is the UAA takes a lot longer to start than ras_backstage, so if we do that then the docker setup it will always fail to start. 

I agree about false impressions but in this scenario it will be obvious. The sign in will fail with an error message in the log